code/modules/reagents/reagent_dispenser.dm

```diff
@@ -118,6 +128,46 @@
 /obj/structure/reagent_dispensers/fueltank/fire_act()
 	blob_act() //saving a few lines of copypasta
 
+/obj/structure/reagent_dispensers/compostbin
+	name = "compost tank"
+	desc = "A device that mulches up unwanted produce into usable fertiliser."
+	icon = 'icons/obj/objects.dmi'
+	icon_state = "comptank"
+	amount_per_transfer_from_this = 30
+	var/list/validCompostTypepaths = list(/obj/item/weapon/reagent_containers/food/snacks/grown, /obj/item/seeds)
+
+/obj/structure/reagent_dispensers/compostbin/attackby(obj/item/weapon/W as obj, mob/user as mob)
+	var/load = 1
+	var/obj/item/weapon/reagent_containers/food/snacks/grown/G = W
+	if(G.compost_value > 0)
+		reagents.add_reagent("compost",G.compost_value)
+	else
+		load = 0
+
+	if(load)
+		user << "<span class='notice'>[src] mulches up [W].</span>"
+		playsound(src.loc, 'sound/effects/blobattack.ogg', 50, 1)
+		user.unEquip(W)
+		qdel(W)
+		return
+	else ..()
+
+/obj/structure/reagent_dispensers/compostbin/MouseDrop_T(atom/movable/O as mob|obj, mob/user as mob)
+	if(is_type_in_list(O, validCompostTypepaths))

```

Comment:
You could check if it's an obj/item and then typecast to obj/item and check the compost value here, instead of doing the istypeinlist

not sure which would be faster.